File: MCN/error detection/hamming.cpp
#include <iostream>

using namespace std;

int main() {
    int d[4], h[7];
    
    cout << "Enter 4 data bits (space separated): ";
    cin >> d[0] >> d[1] >> d[2] >> d[3];

    h[2] = d[0]; 
    h[4] = d[1];
    h[5] = d[2]; 
    h[6] = d[3];

    h[0] = h[2] ^ h[4] ^ h[6];
    h[1] = h[2] ^ h[5] ^ h[6]; 
    h[3] = h[4] ^ h[5] ^ h[6]; 

    cout << "7-bit Hamming Code: ";
    for(int i = 0; i < 7; i++) {
        cout << h[i] << " ";
    }
    cout << endl;

    return 0;
}

File: MCN py/error detection/hamming.py
r = list(map(int, input("Enter 7 received bits (space separated): ").split()))
mode = input("Enter parity mode (even/odd): ").strip().lower()

c1 = r[0] ^ r[2] ^ r[4] ^ r[6]
c2 = r[1] ^ r[2] ^ r[5] ^ r[6]
c4 = r[3] ^ r[4] ^ r[5] ^ r[6]

if mode == 'odd':
    c1 ^= 1
    c2 ^= 1
    c4 ^= 1

error_pos = (c4 * 4) + (c2 * 2) + (c1 * 1)

if error_pos == 0:
    print("Status: No error detected in received data.")
else:
    print(f"Status: Error detected at position {error_pos}.")
    
    r[error_pos - 1] ^= 1 
    print(f"Corrected Hamming Code: {' '.join(map(str, r))}")

print(f"Extracted Data Bits: {r[2]} {r[4]} {r[5]} {r[6]}")